clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -1687,17 +1690,25 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
                     SYCLBufferLocationAttr->getLocationID()))
               : llvm::ConstantAsMetadata::get(CGF->Builder.getInt32(-1)));
 
+      if (parm->hasAttr<RestrictAttr>())
+        argAccessorNoAliasPropertyAttr.push_back(
+            llvm::ConstantAsMetadata::get(CGF->Builder.getInt1(true)));
+
       if (FD->hasAttr<SYCLSimdAttr>())
         argESIMDAccPtrs.push_back(llvm::ConstantAsMetadata::get(
             CGF->Builder.getInt1(parm->hasAttr<SYCLSimdAccessorPtrAttr>())));
     }
 
   bool IsEsimdFunction = FD && FD->hasAttr<SYCLSimdAttr>();
 
-  if (LangOpts.SYCLIsDevice && !IsEsimdFunction)
+  if (LangOpts.SYCLIsDevice && !IsEsimdFunction) {
+
+    Fn->setMetadata("noalias", llvm::MDNode::get(
+                                   VMContext, argAccessorNoAliasPropertyAttr));
+
     Fn->setMetadata("kernel_arg_buffer_location",
                     llvm::MDNode::get(VMContext, argSYCLBufferLocationAttr));
-  else {
+  } else {
```

Comment:
Question to @premanandrao and @DenisBakhvalov : does it meant to be, that kernel_arg_* metadata is being applied to ESIMD functions (aka shall `else` be replaces with `else if`?
Also, I think, that `noalias` function parameter attribute is quiet useful optimization hint and shall be emitted for ESIMD as well.

Reply:
Your suggestion makes sense. I am open to changing it to `else` ; it wasn't clear to me from the spec description about which way to go.

Reply:
> does it meant to be, that kernel_arg_* metadata is being applied to ESIMD functions

I believe it is still a requirement for ESIMD kernels.